src/mjolnir/graphfilter.cc

```diff
@@ -27,10 +29,157 @@ uint32_t n_original_nodes = 0;
+                    const graph_tile_ptr& prev_tile,
+                    GraphId prev_node,
+                    GraphId current_node);
+
+bool ExpandFromNodeInner(GraphReader& reader,
+                         std::list<PointLL>& shape,
+                         GraphId& en,
+                         const GraphId& from_node,
+                         std::unordered_set<std::string>& isos,
+                         bool forward,
+                         GraphId& last_node,
+                         std::unordered_set<GraphId>& visited_nodes,
+                         uint64_t& way_id,
+                         const graph_tile_ptr& prev_tile,
+                         GraphId prev_node,
+                         GraphId current_node,
+                         const NodeInfo* node_info) {
+
+  for (size_t j = 0; j < node_info->edge_count(); ++j) {
+    GraphId edge_id(prev_tile->id().tileid(), prev_tile->id().level(), node_info->edge_index() + j);
+    const DirectedEdge* de = prev_tile->directededge(edge_id);
+    auto edge_info = prev_tile->edgeinfo(de);
+
+    auto tile = prev_tile;
+    if (tile->id() != de->endnode().Tile_Base()) {
+      tile = reader.GetGraphTile(de->endnode());
+    }
+
+    const NodeInfo* en_info = tile->node(de->endnode().id());
+
+    // check the direction, if we looped back, or are we done
+    if ((de->endnode() != prev_node) && (de->forward() == forward) &&
+        (de->endnode() != from_node || (de->endnode() == from_node && visited_nodes.size() > 1))) {
+      if (edge_info.wayid() == way_id &&
+          (en_info->mode_change() || (node_info->mode_change() && !en_info->mode_change()))) {
+
+        // If this edge has special attributes, then we can't aggregate
+        if (!CanAggregate(de)) {
+          way_id = 0;
+          return false;
+        }
+
+        if (isos.size() >= 1) {
+          isos.insert(tile->admin(en_info->admin_index())->country_iso());
+        } else {
+          std::list<PointLL> edgeshape =
+              valhalla::midgard::decode7<std::list<PointLL>>(edge_info.encoded_shape());
+          if (!de->forward()) {
+            std::reverse(edgeshape.begin(), edgeshape.end());
+          }
+
+          // Append shape. Skip first point since it
+          // should equal the last of the prior edge.
+          edgeshape.pop_front();
+          shape.splice(shape.end(), edgeshape);
+        }
+
+        // found a node that does not have aggregation marked (using mode_change flag)
+        // we are done.
+        if (node_info->mode_change() && !en_info->mode_change()) {
+          en = de->endnode();
+          aggregated++;
+          return true;
+        }
+        aggregated++;
+
+        bool found;
+        if (visited_nodes.find(de->endnode()) == visited_nodes.end()) {
+          visited_nodes.insert(de->endnode());
+
+          // expand with the same way_id
+          found = ExpandFromNode(reader, shape, en, from_node, isos, forward, last_node,
```

Comment:
this breaks my head a bit, `ExpandFromNode` is calling `ExpandFromNodeInner` which is calling `ExpandFromNode` which is calling `ExpandFromNodeInner` which is calling :laughing: